add scalar mappings

nikodemin opened the pull request:
Resolves #453

* sealed trait Codegen.GenType instead of passing writer lambda (see Codegen)
* use implicits in SchemaWriter and ClientWriter
* add additional command line option --scalarMappings
* update client.md and schema.md

ghostdogpr commented:
Glad to see this being implemented!

It's a little hard to review because your PR modifies formatting and imports in a bunch of unrelated places ^^ It would be nice if you could clean up so that the PR is only about adding the feature. You can use `fmt` in sbt for formatting.

nikodemin commented:
@ghostdogpr Sorry for inconvenience! I fixed my code, now tests are passing. Also i added one additional feature: additional imports in generated code. I think it will be useful for custom and unsupported types. So you could f.e. define your implicit implementation for ScalarDecoder[OffsetDateTime] and add this import during schema generation. I think this should work when dealing with random type, but I'm not so deep in internals, so please correct me if I'm wrong

ghostdogpr commented:
No problem 😄 sounds good, I'll try to review in details during the weekend!

ghostdogpr commented on tools/src/main/scala/caliban/tools/Codegen.scala:
```diff
@@ -2,24 +2,38 @@ package caliban.tools
 
+  sealed trait GenType
+
+  object Schema extends GenType
```
I usually put these inside `object GenType` so that we can call them `GenType.Schema`

ghostdogpr commented on tools/src/main/scala/caliban/tools/IntrospectionClient.scala:
```diff
@@ -12,8 +12,7 @@ import caliban.parsing.adt.Definition.TypeSystemDefinition.{ DirectiveDefinition
 import sttp.client3._
-import sttp.client3.asynchttpclient.zio.AsyncHttpClientZioBackend
-import sttp.client3.asynchttpclient.zio._
+import sttp.client3.asynchttpclient.zio.{ AsyncHttpClientZioBackend, _ }
```
We can simply remove `AsyncHttpClientZioBackend` I think?

nikodemin replied:
No, it is used on 26 line

ghostdogpr replied:
But since we import `_`, it's not needed. Right?

nikodemin replied:
No, it's used for 29-30 lines. If it was unused, then after optimizing imports it should fade away

ghostdogpr replied:
What I mean is that it's already imported by `_` so there is no need to specify it. Just try to remove it and only keep `_`, you'll see.

nikodemin replied:
Sorry, doesn't noticed that)

ghostdogpr commented on tools/src/main/scala/caliban/tools/SchemaWriter.scala:
```diff
@@ -3,15 +3,17 @@ package caliban.tools
 
   def write(
     schema: Document,
-    objectName: String = "",
```
Nice that we don't have to mirror Client and Schema anymore.

nikodemin replied:
If you mean this writer lambda propagation, like it was before,  I agree)

ghostdogpr commented on tools/src/test/scala/caliban/tools/SchemaWriterSpec.scala:
```diff
@@ -174,9 +183,6 @@ object SchemaWriterSpec extends DefaultRunnableSpec {
           )
         )
       },
-      testM("empty schema test") {
```
Any reason to remove that one?

nikodemin replied:
I removed this test, because on different OS line separators differs, so it's failed on windows. Also I find this test not so useful, but if you want I can return it with corrected line separator

ghostdogpr replied:
Fine with me 👍

nikodemin replied:
@ghostdogpr I returned corrected test

ghostdogpr commented on vuepress/docs/docs/schema.md:
```diff
@@ -242,6 +242,9 @@ The package of the generated code is derived from the folder of `outputPath`. Th
 By default, each Query and Mutation will be wrapped into a `zio.UIO` effect. This can be overridden by providing an alternative effect with the `--effect` option.
 
+If you want to force a mapping between a GraphQL type and a Scala class (such as scalars), you can use the
+`--scalarMappings` option. Also you can add imports for example for your ScalarDecoder implicits by providing `--imports` option.
```
Can you adapt the last sentence? Because there is no `ScalarDecoder` in the server.

ghostdogpr reviewed (commented):
Looks good ! I just have a few minor comments.

nikodemin commented:
@ghostdogpr Thanks for review!
I resolved comments, but there is one open question left about deleted test for empty schema. Please say should I return it or it's redundant

ghostdogpr reviewed (approved):
Looks all good! Thanks for the contribution!
